File: frontend/src/services/course.service.ts
import apiClient, { publicApiClient } from "@/lib/apiClient";

export const courseService = {
  searchCourses: (request: CourseSearchRequest) =>
    publicApiClient.post<CourseSearchApiResponse>(
      "/courses/search",
      request,
    ),
  getCourse: (courseId: string) =>
    publicApiClient.get<ApiResponse<CourseResponse>>(`/courses/${courseId}`),
  getMyCourses: (
    page = 0,
    size = 9,
    keyword = "",
    status?: CourseStatus,
  ) =>
    apiClient.get<ApiResponse<PageResponse<CourseResponse>>>(
      "/courses/my-course",
      { params: { page, size, keyword: keyword || undefined, status } },
    ),
  createCourse: (request: CourseCreationRequest) =>
    apiClient.post<ApiResponse<CourseResponse>>("/courses", request),
  updateCourse: (courseId: string, request: CourseUpdateRequest) =>
    apiClient.put<ApiResponse<CourseResponse>>(`/courses/${courseId}`, request),
  getCurriculum: (courseId: string) =>
    apiClient.get<ApiResponse<CourseCurriculumResponse>>(
      `/courses/${courseId}/curriculum`,
    ),
  getEnrollmentStatus: (courseId: string) =>
    apiClient.get<ApiResponse<CourseEnrollmentStatusResponse>>(
      `/courses/${courseId}/enrollment-status`,
    ),
};

File: backend/src/main/java/com/khangdev/elearningbe/service/course/CourseService.java
package com.khangdev.elearningbe.service.course;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.khangdev.elearningbe.dto.PageResponse;
import com.khangdev.elearningbe.dto.request.course.CourseCreationRequest;
import com.khangdev.elearningbe.dto.request.course.CourseRejectRequest;
import com.khangdev.elearningbe.dto.request.course.CourseSearchRequest;
import com.khangdev.elearningbe.dto.request.course.CourseUpdateRequest;
import com.khangdev.elearningbe.dto.response.course.AdminCourseReviewDetailResponse;
import com.khangdev.elearningbe.dto.response.course.AdminCourseReviewItemResponse;
import com.khangdev.elearningbe.dto.response.course.CoursePublishChecklistResponse;
import com.khangdev.elearningbe.dto.response.course.CourseResponse;
import com.khangdev.elearningbe.dto.response.course.CourseReviewHistoryResponse;
import com.khangdev.elearningbe.enums.CourseStatus;

import java.util.List;
import java.util.UUID;

public interface CourseService {
    CourseResponse createCourse(CourseCreationRequest request);
    CourseResponse updateCourse(UUID courseId, CourseUpdateRequest request);
    PageResponse<CourseResponse> searchCourse(CourseSearchRequest request, int page, int size) throws JsonProcessingException;
    CourseResponse getCourseById(UUID courseId);
    void deleteCourse(UUID courseId);
    PageResponse<CourseResponse> getCourses(UUID instructorId, int page, int size);
    PageResponse<CourseResponse> getCoursesByInstructorUserId(UUID userId, int page, int size, String keyword, CourseStatus status);
    PageResponse<AdminCourseReviewItemResponse> getAdminCourseReviews(
            int page,
            int size,
            String keyword,
            CourseStatus status,
            UUID categoryId,
            String instructor,
            String sortBy
    );
    AdminCourseReviewDetailResponse getAdminCourseReviewDetail(UUID courseId);
    CourseResponse approveCourseReview(UUID courseId);
    CourseResponse rejectCourseReview(UUID courseId, CourseRejectRequest request);
    CoursePublishChecklistResponse getPublishChecklist(UUID courseId);
    List<CourseReviewHistoryResponse> getReviewHistory(UUID courseId);
    CourseResponse submitForReview(UUID courseId);
}
